tests/test_managed_group.rs

```diff
@@ -0,0 +1,341 @@
+            };
+
+        alice_group.process_messages(&queued_messages);
+        bob_group.process_messages(&queued_messages);
+
+        let mut charlie_group = match ManagedGroup::new_from_welcome(
+            &charlie_credential_bundle,
+            &managed_group_config,
+            welcome,
+            Some(bob_group.export_ratchet_tree()),
+            charlie_key_package_bundle,
+        ) {
+            Ok(group) => group,
+            Err(e) => panic!("Error creating group from Welcome: {:?}", e),
+        };
+
+        // Make sure that all groups have the same public tree
+        assert_eq!(
+            alice_group.export_ratchet_tree(),
+            bob_group.export_ratchet_tree(),
+        );
+        assert_eq!(
+            alice_group.export_ratchet_tree(),
+            charlie_group.export_ratchet_tree()
+        );
+
+        // === Charlie sends a message to the group ===
+        let message_charlie = b"Hi, I'm Charlie!";
+        let queued_message =
+            charlie_group.create_message(&charlie_credential_bundle, message_charlie);
+        alice_group.process_messages(&[queued_message.clone()]);
+        bob_group.process_messages(&[queued_message]);
+
+        // === Charlie updates and commits ===
+        let queued_messages = match charlie_group.self_update(&charlie_credential_bundle) {
+            Ok(qm) => qm,
+            Err(e) => panic!("Error performing self-update: {:?}", e),
+        };
+        alice_group.process_messages(&queued_messages);
+        bob_group.process_messages(&queued_messages);
+        charlie_group.process_messages(&queued_messages);
+
+        // Check that all groups have the same state
+        assert_eq!(alice_group.export_secret(&[]), bob_group.export_secret(&[]));
+        assert_eq!(
+            alice_group.export_secret(&[]),
+            charlie_group.export_secret(&[])
+        );
+
+        // Make sure that all groups have the same public tree
+        assert_eq!(
+            alice_group.export_ratchet_tree(),
+            bob_group.export_ratchet_tree(),
+        );
+        assert_eq!(
+            alice_group.export_ratchet_tree(),
+            charlie_group.export_ratchet_tree()
+        );
+
+        // === Charlie removes Bob ===
+        let queued_messages = match charlie_group.remove_members(&charlie_credential_bundle, &[1]) {
+            Ok(qm) => qm,
+            Err(e) => panic!("Could not remove member from group: {:?}", e),
+        };
+        alice_group.process_messages(&queued_messages);
+        bob_group.process_messages(&queued_messages);
+        charlie_group.process_messages(&queued_messages);
+
+        // Make sure that all groups have the same public tree
+        assert_eq!(
+            alice_group.export_ratchet_tree(),
+            bob_group.export_ratchet_tree(),
+        );
+        assert_eq!(
+            alice_group.export_ratchet_tree(),
+            charlie_group.export_ratchet_tree()
+        );
+
```

Comment:
What happens if Bob now tries to send something to the group?

Reply:
Good catch! I forgot to address that. The fix for now is that there is a callback `ErrorOccured` that throws `ManagedGroupError::UseAfterEviction` when this happens. This might not be optimal, the alternative would be to return an error immediately when a group is dead, but it makes the overall API quite ugly.